django/transformer/reference_binder.py

```diff
@@ -0,0 +1,103 @@
+            # (https://www.hl7.org/fhir/references-definitions.html#Reference.reference)
+            try:
+                ref["reference"] = self.identifier_to_reference(identifier, reference_type)
+                logger.debug(
+                    {
+                        "message": f"reference {ref['reference']} of resource {object_id} resolved",
+                        "resource_id": resource_id,
+                    },
+                )
+            except IncompleteIdentifierError as e:
+                logger.warning(
+                    f"incomplete identifier on reference of type "
+                    f"{reference_type} at path {ref} of resource {object_id}: {e}"
+                )
+            except ValueError as e:
+                logger.warning(
+                    f"no valid uuid in identifier on reference of type "
+                    f"{reference_type} at path {ref} of resource {object_id}: {e}"
+                )
+
+        def rec_bind_existing_reference(fhir_object, reference_path: List[str], sub_path=""):
+            if reference_path:
+                # We need to go down the document to find the reference.
+                # reference_path is a list of string so that at the end of each element,
+                # we have an array in the fhir document. For instance, if reference_path
+                # is ["item", "answer.valueReference"], the fhir doc will look like
+                # { item: [ { answer: { valueReference: ... } } ] }.
+                # We want to traverse all the elements of each array that could have
+                # a reference at a leaf.
+                sub_fhir_object = fhir_object[reference_path[0]]
+                sub_path = f"{sub_path}.{reference_path[0]}" if sub_path else reference_path[0]
+                if isinstance(sub_fhir_object, list):
+                    for ind, sub_fhir_el in enumerate(sub_fhir_object):
+                        rec_bind_existing_reference(sub_fhir_el, reference_path[1:], f"{sub_path}.{ind}")
+                else:
+                    rec_bind_existing_reference(sub_fhir_object, reference_path[1:], sub_path)
+            else:
+                # We have isolated the reference so we can now bind it.
+                # If it's an array, we want to perform the binding for all the
+                # elements.
+                if isinstance(fhir_object, list):
+                    for ind, sub_fhir_el in enumerate(fhir_object):
+                        bind(sub_fhir_el)
+                else:
+                    bind(fhir_object)
+
+        rec_bind_existing_reference(fhir_object, reference_path)
+
+    @staticmethod
+    def identifier_to_reference(identifier, reference_type: str) -> str:
+        if not (system := identifier.get("system")):
+            raise IncompleteIdentifierError
+        if not (value := identifier.get("value")):
+            raise IncompleteIdentifierError
+        logical_reference = system[-36:]
```

Comment:
I tend to be afraid of this kind of code.
Here's what make me sleep better:
- use constants like 
 ```
 UUID_LENGTH = 36
...
logical_reference = system[-UUID_LENGTH:]
```
- add comments to explain what's going on (like what a reference.identifier.system should look like)
- or use python's `string` methods like
```python
# system is a string like "http://terminology.arkhn.com/<logical-id-uuid-v4>"
logical_reference = system.split('/')[-1]
```